apis/v1alpha2/httproute_types.go

```diff
@@ -0,0 +1,752 @@
+	//     values:
+	//       version: "2"
+	// - path:
+	//     value: "/v2/foo"
+	// ```
+	//
+	// For a request to match against this rule, a request should satisfy
+	// EITHER of the two conditions:
+	//
+	// - path prefixed with `/foo` AND contains the header `version: "2"`
+	// - path prefix of `/v2/foo`
+	//
+	// See the documentation for HTTPRouteMatch on how to specify multiple
+	// match conditions that should be ANDed together.
+	//
+	// If no matches are specified, the default is a prefix
+	// path match on "/", which has the effect of matching every
+	// HTTP request.
+	//
+	//
+	// Each client request MUST map to a maximum of one route rule. If a request
+	// matches multiple rules, matching precedence MUST be determined in order
+	// of the following criteria, continuing on ties:
+	//
+	// * The longest matching hostname.
+	// * The longest matching path.
+	// * The largest number of header matches.
+	//
+	// If ties still exist across multiple Routes, matching precedence MUST be
+	// determined in order of the following criteria, continuing on ties:
+	//
+	// * The oldest Route based on creation timestamp. For example, a Route with
+	//   a creation timestamp of "2020-09-08 01:02:03" is given precedence over
+	//   a Route with a creation timestamp of "2020-09-08 01:02:04".
+	// * The Route appearing first in alphabetical order by
+	//   "<namespace>/<name>". For example, foo/bar is given precedence over
+	//   foo/baz.
+	//
+	// If ties still exist within the Route that has been given precedence,
+	// matching precedence MUST be granted to the first matching rule meeting
+	// the above criteria.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	// +kubebuilder:default={{path:{ type: "Prefix", value: "/"}}}
+	Matches []HTTPRouteMatch `json:"matches,omitempty"`
+
+	// Filters define the filters that are applied to requests that match
+	// this rule.
+	//
+	// The effects of ordering of multiple behaviors are currently unspecified.
+	// This can change in the future based on feedback during the alpha stage.
+	//
+	// Conformance-levels at this level are defined based on the type of filter:
+	//
+	// - ALL core filters MUST be supported by all implementations.
+	// - Implementers are encouraged to support extended filters.
+	// - Implementation-specific custom filters have no API guarantees across
+	//   implementations.
+	//
+	// Specifying a core filter multiple times has unspecified or custom conformance.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Filters []HTTPRouteFilter `json:"filters,omitempty"`
+
+	// BackendRefs defines the backend(s) where matching requests should be
+	// sent. If unspecified or invalid (refers to a non-existent resource or a Service with no endpoints),
+	// the rule performs no forwarding; if no filters are specified that would result in a
+	// response being sent, a HTTP 503 status code is returned. 503 responses must be sent so that the overall
```

Comment:
Redirects are an example of a filter that would cause `BackendRefs` to never be reached. Other examples include an authorization filter that could result in a 401 being returned before we reached `BackendRefs`.

Reply:
right, I get it in the case where there _are_ filters, but the text says that if no filters are selected, then you get a 503, whereas I would have assumed that if no filters are selected, it would just go directly to the BackendRefs?

Reply:
oh yep, that's confusing. This was intended to be a follow up from the previous sentence of "if there are no valid backends", restructuring this in #796 to make it clearer.